### PR TITLE
feat(hooks): implement "classNames" prop in all components

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

Implements classNames in all hooks components, like in #3379 

**Result**

- each component accepts classNames
- global tests for "common behaviour" between all components

### CODE REVIEW

**Review by sarahdayan — COMMENTED** [review]
As brought up in the [`<PoweredBy>` PR](https://github.com/algolia/react-instantsearch/pull/3391), I'm not sure passing the default classes from the widgets is the right call.

The class names are part of the UI spec, so I believe it makes sense to provide them with the components. This way, we don't have to repeat the same logic across every flavor of InstantSearch, which minimizes the risks of inconsistencies.

Example of what it looks like:

- [`<PoweredBy>` component](https://github.com/algolia/react-instantsearch/blob/06ac1765dde2e3fb7b91bd61948e01976f846947/packages/react-instantsearch-hooks-dom/src/ui/PoweredBy.tsx#L21)
- [`<PoweredBy>` widget](https://github.com/algolia/react-instantsearch/blob/06ac1765dde2e3fb7b91bd61948e01976f846947/packages/react-instantsearch-hooks-dom/src/widgets/PoweredBy.tsx#L11-L15)

File: packages/react-instantsearch-hooks-dom/src/ui/SearchBox.tsx [hunk 1]
```diff
@@ -17,15 +17,59 @@ export type SearchBoxProps = React.HTMLAttributes<HTMLDivElement> & {
...
+   */
+  submitIcon: string;
+  /**
+   * Class names to apply to the reset icon.
+   */
+  resetIcon: string;
+  /**
+   * Class names to apply to the loading icon.
+   */
+  loadingIcon: string;
```
Comment (dhayab) [inline packages/react-instantsearch-hooks-dom/src/ui/SearchBox.tsx:66]:
```suggestion
  /**
   * Class names to apply to the root element
   */
  root: string;
  /**
   * Class names to apply to the form element
   */
  form: string;
  /**
   * Class names to apply to the input element
   */
  input: string;
  /**
   * Class names to apply to the submit button
   */
  submit: string;
  /**
   * Class names to apply to the reset button
   */
  reset: string;
  /**
   * Class names to apply to the loading indicator element
   */
  loadingIndicator: string;
  /**
   * Class names to apply to the submit icon
   */
  submitIcon: string;
  /**
   * Class names to apply to the reset icon
   */
  resetIcon: string;
  /**
   * Class names to apply to the loading icon
   */
  loadingIcon: string;
```

**Review by dhayab — COMMENTED** [review]
Small suggestion to normalize prop descriptions

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/CurrentRefinements.test.tsx [hunk 2]
```diff
@@ -621,6 +621,39 @@ describe('CurrentRefinements', () => {
...
+          classNames={{
+            root: 'ROOT',
+            rootNoRefinement: 'ROOTNOREFINEMENT',
+            list: 'LIST',
+            listNoRefinement: 'LISTNOREFINEMENT',
+            item: 'ITEM',
+            label: 'LABEL',
+            category: 'CATEGORY',
+            categoryLabel: 'CATEGORYLABEL',
+            delete: 'DELETE',
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/__tests__/CurrentRefinements.test.tsx:638]: These aren't tested.

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/Hits.test.tsx [hunk 3]
```diff
@@ -111,6 +111,33 @@ describe('Hits', () => {
...
 
+  test('accepts custom class names', () => {
+    const { container } = render(
+      <InstantSearchHooksTestWrapper>
+        <Hits
+          className="MyHits"
+          classNames={{
+            root: 'ROOT',
+            list: 'LIST',
+            item: 'ITEM',
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/__tests__/Hits.test.tsx:122]: This isn't tested.

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-widgets.test.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
+/**
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-widgets.test.tsx:1]: This refactor seems unrelated to this PR.

File: packages/react-instantsearch-hooks-dom/src/widgets/__testutils__/all-widgets.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
+import React from 'react';
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/__testutils__/all-widgets.tsx:1]: This refactor seems unrelated to this PR.

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-components.test.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/**
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-components.test.tsx:1]: Do we need this test file? We're testing this in each individual widget already.

File: packages/react-instantsearch-hooks-dom/src/widgets/Hits.tsx [hunk 7]
```diff
@@ -13,7 +13,9 @@ export type HitsProps<THit extends BaseHit> = Omit<
 > &
   UseHitsProps<THit>;
 
-export function Hits<THit extends BaseHit = BaseHit>(props: HitsProps<THit>) {
+export function Hits<THit extends BaseHit = BaseHit>({
+  ...props
```
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/widgets/Hits.tsx:17]: Why this change?

File: packages/react-instantsearch-hooks-dom/src/ui/SearchBox.tsx (hunk 1 above)
Comment (sarahdayan) [inline packages/react-instantsearch-hooks-dom/src/ui/SearchBox.tsx:66]:
```suggestion
  /**
   * Class names to apply to the root element
   */
  root: string;
  /**
   * Class names to apply to the form element
   */
  form: string;
  /**
   * Class names to apply to the input element
   */
  input: string;
  /**
   * Class names to apply to the submit button
   */
  submit: string;
  /**
   * Class names to apply to the reset button
   */
  reset: string;
  /**
   * Class names to apply to the loading indicator element
   */
  loadingIndicator: string;
  /**
   * Class names to apply to the submit icon
   */
  submitIcon: string;
  /**
   * Class names to apply to the reset icon
   */
  resetIcon: string;
  /**
   * Class names to apply to the loading icon
   */
  loadingIcon: string;
```

File: packages/react-instantsearch-hooks-dom/src/ui/Hits.tsx [hunk 8]
```diff
@@ -17,16 +18,38 @@ function DefaultHitComponent({ hit }: { hit: Hit }) {
   );
 }
 
+export type HitsClassNames = {
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply packages/react-instantsearch-hooks-dom/src/ui/Hits.tsx:21]: I think we just forgot, will follow up afterwards

File: packages/react-instantsearch-hooks-dom/src/widgets/Hits.tsx (hunk 7 above)
Comment (Haroenv) [reply packages/react-instantsearch-hooks-dom/src/widgets/Hits.tsx:17]: needed it for classnames at widget level at some point, no longer now so can indeed be reverted

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-components.test.tsx (hunk 6 above)
Comment (Haroenv) [reply packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-components.test.tsx:1]: this is making sure we never forget it in a test. See it as contract-testing

File: packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-widgets.test.tsx (hunk 4 above)
Comment (Haroenv) [reply packages/react-instantsearch-hooks-dom/src/widgets/__tests__/all-widgets.test.tsx:1]: no, it's required, otherwise "all-components.test.ts" can't be written (jsdom is required for testing library)

**Comment by sarahdayan** [comment]
`<RefinementList>` was merged in the meantime, do you want to rebase to update it?

**Review by sarahdayan — APPROVED** [review]

**Review by dhayab — APPROVED** [review]
That's great, thank you!
